docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+- Verifier application
+    1. Upon initialization, queries available verifiers to the application (oe_sgx_get_verifiers) and registers them (oe_register_verifier).
+    2. Builds the format IDs of the registered verifiers sorted by preference.
+    3. Sends the format IDs to the attester application. 
+- Attester application
+    1. Upon initialization, queries available attesters to the application (oe_sgx_get_attesters) and registers them (oe_register_attester).
+    2. Receives format IDs from the verifier application.
+    3. Selects a format ID from the received format IDs (oe_select_attester_format).
+    4. Sends the format ID to the verifier application.
+- Verifier application
+    1. Gets evidence parameters for the received format ID (oe_get_evidence_parameters).
+    2. Sends the evidence parameters to the attester application.
+- Attester application
+    1. Generates an evidence (oe_get_evidence) of the selected format with the received evidence parameters.
+    2. Sends the evidence to the verifier application.
+- Verifier application
+    1. Verifies the received evidence (oe_verify_evidence).
+
+Scenario 2 - Attester sends evidences of different formats until the verifier can verify:
+- Attester application
+    1. Upon initialization, queries available attesters to the application (oe_sgx_get_attesters) and registers them (oe_register_attester).
+    2. Builds the format IDs of registered attester sorted by perference.
+    3. Generates an evidence (oe_get_evidence) of the first format ID from the list built in step 2.
+    4. Sends the evidence to the verifier application.
+- Verifier application
+    1. Upon initialization, queries available verifiers to the application (oe_sgx_get_verifiers) and registers them (oe_register_verifier).
+    2. Verifies the received evidence (oe_verify_evidence). If the verification fails, ask the attester applicaiton to generate an evidence of different format.
+- Attester application
+    1. Generates an evidence (oe_get_evidence) of the next format from registered attesters and sends it to the verifier application.
+    2. Repeat until all formats are attempted or verification succeeds.
+
+Specification
+-------------
+
+## Differences between V2 and V3 APIs.
+
+Most of the APIs to attestation plugins are the same between the versions. The differences are highlighted below.
+
+## Changes
+
+### New APIs to query and select attesters and verifiers
+
+```C
+/**
+ * oe_sgx_get_attesters
+ * 
+ * Helper function that returns the SGX attesters that can then be sent to
+ * `oe_register_attester`.
+ *
+ * @param[out] attesters The SGX attesters that are available to the application.
+ * @param[out] attesters_length The length of the SGX attesters.
+ * @retval OE_OK on success.
+ * @retval OE_INVALID_PARAMETER At least one parameter is invalid.
+ * @retval other appropriate error code.
+ */
+oe_result_t oe_sgx_get_attesters(oe_attester_t** attesters, size_t *attesters_length);
```

Comment:
As noted above, I haven't seen sufficient justification in this document to warrant any API with _sgx_ in the name.

Reply:
This is a replacement of existing API that is to be deprecated:
oe_sgx_plugin_attester()

---
In reply to: [393256309](https://github.com/openenclave/openenclave/pull/2621#discussion_r393256309) [](ancestors = 393256309)

Reply:
As noted above, I don't consider "because a bad API we want to deprecate did it" to be "sufficient justification".

Reply:
Rename
. oe_sgx_get_attesters to oe_get_attesters
. oe_sgx_get_verifiers to oe_get_verifiers

---
In reply to: [394656586](https://github.com/openenclave/openenclave/pull/2621#discussion_r394656586) [](ancestors = 394656586)